### PR TITLE
Proposal SciML+NeuralClosure

### PR DESCRIPTION
**Pull request by luisaforozco**
In this PR @v1kko and me have worked on a set of functions to handle the data and interface with InompressibleNavierStokes.jl
We propose the workflow for generating data and training via apriori and a-posteriori fitting in the script [simulations/NavierStokes_2D/scripts/NeuralClosure+SciML.jl](https://github.com/DEEPDIP-project/CoupledNODE.jl/blob/luisaforozco/proposal_clean/simulations/NavierStokes_2D/scripts/NeuralClosure%2BSciML.jl).
We've also added a module `NavierStokes` that contains utilities necessary for this purpose.

Not much clean-up (deletion) of files has been done in this PR.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -60,11 +62,11 @@ ParameterSchedulers = "d7d3b36b-41b8-4d0d-a2bf-768c6151755e"
 Plotly = "58dd65bb-95f3-509e-9936-c39a10fdeae7"
 Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
 PreallocationTools = "d236fae5-4411-538c-8e31-a6e3d9e00b46"
-ProfileView = "c46f51b8-102a-5cf2-8d2c-8597cb0e0da7"
 PyPlot = "d330b81b-6aea-500a-939a-2ce795aea3ee"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 ReadVTK = "dc215faf-f008-4882-a9f7-a79a826fadc3"
 RecursiveArrayTools = "731186ca-8d62-57ce-b412-fbd966d074cd"
+Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
```
Comment (v1kko) [inline Project.toml:69]:
```suggestion
```
The code works fine without Revise

File: Project.toml [hunk 2]
```diff
@@ -31,8 +31,9 @@ Interpolations = "a98d9a8b-a2ab-59e6-89dd-64a1c18fca59"
 IterativeSolvers = "42fd0dbc-a981-5370-80f2-aaf504508153"
 JLD2 = "033835bb-8acc-5ee8-8aae-3f567f8a3819"
 JuliaFormatter = "98e50ef6-434e-11e9-1051-2b60c6c9e899"
-Kronecker = "2c470bb0-bcc8-11e8-3dad-c9649493f05e"
+Juno = "e5e0dc1b-0480-54bc-9374-aad01c23163d"
 KernelAbstractions = "63c18a36-062a-441e-b654-da1e3ab1ce7c"
+Kronecker = "2c470bb0-bcc8-11e8-3dad-c9649493f05e"
```
Comment (v1kko) [inline Project.toml:36]:
```suggestion
```
Is this package used? I do not see it in the code

File: Project.toml (hunk 1 above)
Comment (luisaforozco) [reply Project.toml:69]: yes, but I think we want to have Revise...

File: Project.toml (hunk 2 above)
Comment (luisaforozco) [reply Project.toml:36]: It is used in test_derivatives. Going through the dependencies and cleaning up what we do not need would be next after agreeing that we want this PR merged and using the workflow in the script.

Comment (SCiarella) [reply Project.toml:36]: I suggest that later after the cleanup we start using [`[weakdeps]` and `[extensions]`](https://pkgdocs.julialang.org/v1/creating-packages/#Weak-dependencies)

Comment (luisaforozco) [reply Project.toml:36]: Sure! nice to check what can be the best practices there.

File: Project.toml (hunk 1 above)
Comment (v1kko) [reply Project.toml:69]: Ok, its not so important now, lets look at it later

File: Project.toml (hunk 2 above)
Comment (v1kko) [reply Project.toml:36]: I added the link to the best practices in the relevant issue. lets do this later then

File: src/equations/NavierStokes_utils.jl [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+# Returns
+- A function `dataloader` that, when called, returns a tuple with:
+  - `u`: A view into the `u` array of `io_array` for a randomly selected sample and time steps.
+  - `t`: The corresponding time steps from the `t` array of `io_array`.
+
+# Notes
+- The `io_array.u` array is expected to have dimensions `(n, n, dim, samples, nt)`.
+- The `nt` dimension must be greater than or equal to `nunroll`.
+- Have only tested in 2D cases.
```
Comment (SCiarella) [inline src/equations/NavierStokes_utils.jl:305]:
```suggestion
- Have only tested in 2D cases.
- It assumes that the data are loaded in batches of size 1
```

File: src/equations/NavierStokes_utils.jl [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+"""
+    create_right_hand_side_with_closure(setup, psolver, closure, st)
+
+Create right hand side function f(u, p, t) compatible with SciML ODEProblem.
+This formulation was the one proved best in Syver's paper i.e. DCF.
+`u` has to be an array in the NN style e.g. `[n , n, D]`, with boundary conditions padding.
+"""
+create_right_hand_side_with_closure(setup, psolver, closure, st) = function right_hand_side(
+        u, p, t)
+    # not sure if we should keep t as a parameter. t is only necessary for the INS functions when having Dirichlet BCs (time dependent)
```
Comment (SCiarella) [inline src/equations/NavierStokes_utils.jl:37]: SciML expects functions to have signature `f(u, p, t)` so I would leave it like this even if we do not use it

File: src/equations/NavierStokes_utils.jl [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+to a format suitable for neural network training `u[n, n, D, batch]`.
+
+# Arguments
+- `u`: Velocity field in INS style.
+
+# Returns
+- `u`: Velocity field converted to a tensor format suitable for neural network training.
+"""
+function INS_to_NN(u)
+    cat(u...; dims = ndims(u[1]) + 1)
```
Comment (SCiarella) [inline src/equations/NavierStokes_utils.jl:140]: why not this to avoid copy and allocation
```suggestion
    view(u, 1)
```

File: src/loss/loss_posteriori.jl [hunk 6]
```diff
@@ -146,3 +147,48 @@ function loss_MulDtO_oneset(trajectory,
 
     return loss + (continuity * λ_c) + λ_l1 * norm(θ), nothing
 end
+
+"""
+    create_loss_post_lux(rhs; sciml_solver = Tsit5(), kwargs...)
+
+Creates a loss function for a-posteriori fitting using the given right-hand side (RHS) function `rhs`.
+The loss function computes the sum of squared differences between the predicted values (autoregressively) and the actual data,
```
Comment (SCiarella) [inline src/loss/loss_posteriori.jl:155]:
```suggestion
The loss function computes the sum of squared differences between the predicted values and the actual data, 
```

File: src/loss/loss_posteriori.jl [hunk 7]
```diff
@@ -146,3 +147,48 @@ function loss_MulDtO_oneset(trajectory,
...
+- `metadata::NamedTuple`: A named tuple containing the predicted values `y_pred`.
+This makes it compatible with the Lux ecosystem.
+"""
+function create_loss_post_lux(rhs; sciml_solver = Tsit5(), kwargs...)
+    function loss_function(model, ps, st, (u, t))
+        griddims = Zygote.@ignore ((:) for _ in 1:(ndims(u) - 2))
+        x = u[griddims..., :, 1]
+        y = u[griddims..., :, 2:end] # remember to discard sol at the initial time step
+        dt = t[2] - t[1]
+        #saveat_loss = [i * dt for i in 1:length(y)]
```
Comment (SCiarella) [inline src/loss/loss_posteriori.jl:184]: this function is not accounting for the fact that one may want to measure the loss using a sampling time $>dt$, which can be easily done using saveat_loss. For now we can ignore this, but let's keep in mind that now we are forced to very narrow t-intervals for the a-posteriori loss

**Review by SCiarella — APPROVED** [review]
Good job guys! :rocket: 

I have left some comments, which are mainly minor edits.
The only major issue that I see is that I am not convinced that the a priori training with Lux is working as expected.

Anyway this PR can be already merged for me :+1:

File: src/loss/loss_posteriori.jl (hunk 7 above)
Comment (luisaforozco) [reply src/loss/loss_posteriori.jl:184]: Indeed, I kept it simple for the sake of understanding the code. Feel free to open an issue or propose a change here.

File: src/loss/loss_priori.jl [hunk 8]
```diff
@@ -98,3 +96,48 @@ function create_randloss_derivative(
...
+
+"""
+Compute MSE between `f(x, θ, st)` and `y`.
+The MSE is further divided by `normalize(y)`.
+This signature has been chosen for compatibility with Lux.
+"""
+mean_squared_error(f, θ, st, (x, y); normalize = y -> sum(abs2, y), λ = sqrt(1e-8)) = sum(
+    abs2, f(x, θ, st)[1] - y) / normalize(y) + eltype(x)(λ) *
+                                                                                      sum(
+    abs2, θ)
```
[the remark this answers is not in the document]
Comment (luisaforozco) [reply src/loss/loss_priori.jl:120]: Yes, is still being used to create `loss_priori` (used for option 1 of training). I did however delete the `create_loss_priori` that would create a `Lux.GenericLossFunction` because that was not correct nor used.

File: src/equations/NavierStokes_utils.jl (hunk 5 above)
Comment (luisaforozco) [reply src/equations/NavierStokes_utils.jl:140]: @v1kko do you have an opinion about this?

File: simulations/NavierStokes_2D/scripts/NeuralClosure+SciML.jl [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+# * new way of training (via Lux)
+using CoupledNODE: train
+loss, tstate = train(closure, θ, st, dataloader_prior, loss_priori_lux;
+    nepochs = 50, ad_type = Optimization.AutoZygote(),
+    alg = OptimizationOptimisers.Adam(0.1), cpu = true, callback = callback_2)
+# the trained parameters at the end of the training are: 
+θ_priori = tstate.parameters
+# and the best parameters found during training are:
+θ_priori_best = callbackstate.θmin
```
[the remark this answers is not in the document]
Comment (luisaforozco) [reply simulations/NavierStokes_2D/scripts/NeuralClosure+SciML.jl:103]: @SCiarella is going to explore this: difference between the training via Optimization.solve and Lux

File: src/equations/NavierStokes_utils.jl (hunk 5 above)
Comment (v1kko) [reply src/equations/NavierStokes_utils.jl:140]: It looks good, I say use it if it works

Comment (luisaforozco) [reply src/equations/NavierStokes_utils.jl:140]: No, it doesn't work. Indeed u is a Tuple, and there is now `view` function typed for `Tuple`. Here the code that you can try to see that it doesn't work:
```julia
u0 = INS.random_field(setups[1], 0)
op1 = cat(u0...; dims = ndims(u0[1]) + 1) # current, working
op2 = view(u0, 1) # proposed, doesn't work
```
